src/Particles/babylon.solidParticleSystem.ts

```diff
@@ -0,0 +1,427 @@
+
+                    if (this._useParticleColor) {
+                        this._colors[colidx] = this._particle.color.r;
+                        this._colors[colidx + 1] = this._particle.color.g;
+                        this._colors[colidx + 2] = this._particle.color.b;
+                        this._colors[colidx + 3] = this._particle.color.a;
+                    }
+
+                    if (this._useParticleTexture) {
+                        this._uvs[uvidx] = this._particle._shapeUV[pt * 2] * (this._particle.uvs[2] - this._particle.uvs[0]) + this._particle.uvs[0];
+                        this._uvs[uvidx + 1] = this._particle._shapeUV[pt * 2 + 1] * (this._particle.uvs[3] - this._particle.uvs[1]) + this._particle.uvs[1];
+                    }
+                }
+                index = idx + 3;
+                colorIndex = colidx + 4;
+                uvIndex = uvidx + 2;
+            }
+
+            if (this._useParticleColor) {
+                this.mesh.updateVerticesData(VertexBuffer.ColorKind, this._colors, false, false);
+            }
+            if (this._useParticleTexture) {
+                this.mesh.updateVerticesData(VertexBuffer.UVKind, this._uvs, false, false);
+            }
+            this.mesh.updateVerticesData(VertexBuffer.PositionKind, this._positions, false, false);
+            if (!this.mesh.areNormalsFrozen) {
+                var indices = this.mesh.getIndices();
+                VertexData.ComputeNormals(this._positions, this._indices, this._normals);
+                this.mesh.updateVerticesData(VertexBuffer.NormalKind, this._normals, false, false);
+            }
+            this.afterUpdateParticles();
+        }
+        // internal implementation of BJS Quaternion.RotationYawPitchRollToRef() with memory reuse
+        private _quaternionRotationYPR(): void {
+            this._halfroll = this._roll * 0.5;
+            this._halfpitch = this._pitch * 0.5;
+            this._halfyaw = this._yaw * 0.5;
+            this._sinRoll = Math.sin(this._halfroll);
+            this._cosRoll = Math.cos(this._halfroll);
+            this._sinPitch = Math.sin(this._halfpitch);
+            this._cosPitch = Math.cos(this._halfpitch);
+            this._sinYaw = Math.sin(this._halfyaw);
+            this._cosYaw = Math.cos(this._halfyaw);
+            this._quaternion.x = (this._cosYaw * this._sinPitch * this._cosRoll) + (this._sinYaw * this._cosPitch * this._sinRoll);
+            this._quaternion.y = (this._sinYaw * this._cosPitch * this._cosRoll) - (this._cosYaw * this._sinPitch * this._sinRoll);
+            this._quaternion.z = (this._cosYaw * this._cosPitch * this._sinRoll) - (this._sinYaw * this._sinPitch * this._cosRoll);
+            this._quaternion.w = (this._cosYaw * this._cosPitch * this._cosRoll) + (this._sinYaw * this._sinPitch * this._sinRoll);
+        }
+
+        // internal implemenation of BJS toRotationMatric() with memory reuse
+        private _quaternionToRotationMatrix(): void {
```

Comment:
because of the garbage collection : these reimplementation just 
allocates/reallocates only 8 variables
these 8 variables belong to the SPS object
as we iterate over thousands of objects, using the core math functions, 
which create new temp variable each call, will crumble the FPS
this has been over-tested : there's a huge gain to reimplement these 
dedicated methods here


Le 14/10/2015 17:14, David Catuhe a écrit :
>
> In src/Particles/babylon.solidParticleSystem.ts 
> <https://github.com/BabylonJS/Babylon.js/pull/751#discussion_r42006866>:
>
> > +            this._halfpitch = this._pitch * 0.5;
> > +            this._halfyaw = this._yaw * 0.5;
> > +            this._sinRoll = Math.sin(this._halfroll);
> > +            this._cosRoll = Math.cos(this._halfroll);
> > +            this._sinPitch = Math.sin(this._halfpitch);
> > +            this._cosPitch = Math.cos(this._halfpitch);
> > +            this._sinYaw = Math.sin(this._halfyaw);
> > +            this._cosYaw = Math.cos(this._halfyaw);
> > +            this._quaternion.x = (this._cosYaw * this._sinPitch * this._cosRoll) + (this._sinYaw * this._cosPitch * this._sinRoll);
> > +            this._quaternion.y = (this._sinYaw * this._cosPitch * this._cosRoll) - (this._cosYaw * this._sinPitch * this._sinRoll);
> > +            this._quaternion.z = (this._cosYaw * this._cosPitch * this._sinRoll) - (this._sinYaw * this._sinPitch * this._cosRoll);
> > +            this._quaternion.w = (this._cosYaw * this._cosPitch * this._cosRoll) + (this._sinYaw * this._sinPitch * this._sinRoll);
> > +        }
> > +
> > +        // internal implemenation of BJS toRotationMatric() with memory reuse
> > +        private _quaternionToRotationMatrix(): void {
>
> Why not in math.ts?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/BabylonJS/Babylon.js/pull/751/files#r42006866>.
>